support/agent/miner_agent.go

```diff
@@ -156,34 +195,187 @@ func (ma *MinerAgent) Tick(v SimState) ([]message, error) {
+
+	return message{
+		From:   ma.Worker,
+		To:     ma.IDAddress,
+		Value:  big.Zero(),
+		Method: builtin.MethodsMiner.ProveCommitSector,
+		Params: &params,
+	}
+}
+
+// Fault a sector.
+// This chooses a sector from live sectors and then either declares the recovery
+// or adds it as a fault
+func (ma *MinerAgent) createFault(v SimState) ([]message, error) {
+	// opt out if no live sectors
+	if len(ma.liveSectors) == 0 {
+		return nil, nil
+	}
+
+	// choose a live sector to go faulty
+	var faultNumber uint64
+	faultNumber, ma.liveSectors = PopRandom(ma.liveSectors, ma.rnd)
+	ma.faultySectors = append(ma.faultySectors, faultNumber)
+
+	faultDlInfo, pIdx, err := ma.dlInfoForSector(v, faultNumber)
+	if err != nil {
+		return nil, err
+	}
+
+	parts := ma.deadlines[faultDlInfo.Index]
+	if pIdx >= uint64(len(parts)) {
+		return nil, errors.Errorf("sector %d in deadline %d has unregistered partition %d",
+			faultNumber, faultDlInfo.Index, pIdx)
+	}
+	parts[pIdx].faults.Set(faultNumber)
+
+	// If it's too late, skip fault rather than declaring it
+	if faultDlInfo.FaultCutoffPassed() {
+		parts[pIdx].toBeSkipped.Set(faultNumber)
+		return nil, nil
+	}
+
+	// for now, just send a message per fault rather than trying to batch them
+	faultParams := miner.DeclareFaultsParams{
+		Faults: []miner.FaultDeclaration{{
+			Deadline:  faultDlInfo.Index,
+			Partition: pIdx,
+			Sectors:   bitfield.NewFromSet([]uint64{faultNumber}),
+		}},
+	}
+
+	return []message{{
+		From:   ma.Worker,
+		To:     ma.IDAddress,
+		Value:  big.Zero(),
+		Method: builtin.MethodsMiner.DeclareFaults,
+		Params: &faultParams,
+	}}, nil
+}
+
+// Recover a sector.
+// This chooses a sector from faulty sectors and then either declare the recovery or schedule one for later
+func (ma *MinerAgent) createRecovery(v SimState) ([]message, error) {
+	// opt out if no faulty sectors
+	if len(ma.faultySectors) == 0 {
+		return nil, nil
+	}
+
+	// choose a faulty sector to recover
+	var recoveryNumber uint64
+	recoveryNumber, ma.faultySectors = PopRandom(ma.faultySectors, ma.rnd)
+
+	recoveryDlInfo, pIdx, err := ma.dlInfoForSector(v, recoveryNumber)
+	if err != nil {
+		return nil, err
+	}
+
+	parts := ma.deadlines[recoveryDlInfo.Index]
+	if pIdx >= uint64(len(parts)) {
+		return nil, errors.Errorf("recovered sector %d in deadline %d has unregistered partition %d",
+			recoveryNumber, recoveryDlInfo.Index, pIdx)
+	}
+	if set, err := parts[pIdx].faults.IsSet(recoveryNumber); err != nil {
+		return nil, errors.Errorf("could not check if %d in deadline %d partition %d is faulty",
+			recoveryNumber, recoveryDlInfo.Index, pIdx)
+	} else if !set {
+		return nil, errors.Errorf("recovery %d in deadline %d partition %d was not a fault",
+			recoveryNumber, recoveryDlInfo.Index, pIdx)
+	}
+
+	// If it's too late, schedule recovery rather than declaring it
+	if recoveryDlInfo.FaultCutoffPassed() {
+		ma.operationSchedule.ScheduleOp(recoveryDlInfo.Close, recoverSectorAction{
```

Comment:
Out of curiosity do we go through the operation schedule when we terminate or expire a sector to remove all outstanding operations on terminated sectors?

Reply:
we do now.

Reply:
We don't, but only `registerSector` and `delayedRecoveryMessage` operate on individual on-chain sectors and they both handle missing sectors gracefully.